### PR TITLE
Coldstart optimizations (#5679, #5704)

### PR DESCRIPTION
**Pull request by mathewc**
Addresses https://github.com/Azure/azure-functions-host/issues/5679 and https://github.com/Azure/azure-functions-host/issues/5704.

This work is in a similar vein to what @ankitkumarr did recently in https://github.com/Azure/azure-functions-host/pull/5607.

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Security/KeyManagement/BaseSecretsRepository.cs [hunk 1]
```diff
@@ -14,30 +13,37 @@ namespace Microsoft.Azure.WebJobs.Script.WebHost
 {
     public abstract class BaseSecretsRepository : ISecretsRepository, IDisposable
     {
-        private readonly AutoRecoveringFileSystemWatcher _sentinelFileWatcher;
+        private const int SentinelWatcherInitializationDelayMS = 3 * 1000;
```
Comment (safihamid) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/BaseSecretsRepository.cs]: Can we make this 5 sec as our upper bound for cold start when extra dependencies are brought by function code? Also this only is needed for Consumption plan. So no need to delay this for AppService plan or for the warm instances of premium if we can detect that.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/BlobStorageSecretsRepository.cs [hunk 2]
```diff
@@ -47,9 +42,19 @@ public BlobStorageSecretsRepository(string secretSentinelDirectoryPath, string a
 
             _secretsBlobPath = siteSlotName.ToLowerInvariant();
             _hostSecretsBlobPath = string.Format("{0}/{1}", _secretsBlobPath, ScriptConstants.HostMetadataFileName);
-
             _accountConnectionString = accountConnectionString;
-            _blobContainer = CreateBlobContainer(_accountConnectionString);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/BlobStorageSecretsRepository.cs]: Shouldn't access this on construction. Instead create on demand.

File: src/WebJobs.Script.WebHost/Security/KeyManagement/BlobStorageSecretsRepository.cs [hunk 3]
```diff
@@ -62,12 +67,11 @@ public override bool IsEncryptionSupported
...
         {
-            CloudStorageAccount account = CloudStorageAccount.Parse(_accountConnectionString);
-            CloudBlobClient client = account.CreateCloudBlobClient();
-            CloudBlobContainer container = client.GetContainerReference(_secretsContainerName);
+            var account = CloudStorageAccount.Parse(connectionString);
+            var client = account.CreateCloudBlobClient();
+            var container = client.GetContainerReference(_secretsContainerName);
 
-            // TODO: Remove this (it is already slated to be removed)
-            container.CreateIfNotExistsAsync().GetAwaiter().GetResult();
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Security/KeyManagement/BlobStorageSecretsRepository.cs]: not sure why we were doing this unecessary sync over async

File: src/WebJobs.Script/Utility.cs [hunk 4]
```diff
@@ -692,6 +694,19 @@ public static void ExecuteAfterDelay(Action targetAction, TimeSpan delay, Cancel
             }, TaskContinuationOptions.OnlyOnRanToCompletion);
         }
 
+        public static void ExecuteWithColdstartDelay(IEnvironment environment, Action targetAction, CancellationToken cancellationToken = default)
```
Comment (mathewc) [inline src/WebJobs.Script/Utility.cs]: Creating a single explicitly named helper that we can use for all these coldstart delays, so we're handling it consistently everywhere.

Comment (safihamid) [reply src/WebJobs.Script/Utility.cs]: Great! Shouldn't it be ExecuteWithColdStartDelay? :-)

---
In reply to: [384734964](https://github.com/Azure/azure-functions-host/pull/5697#discussion_r384734964) [](ancestors = 384734964)

Comment (mathewc) [reply src/WebJobs.Script/Utility.cs]: Renamed

**Pull request merged**